### PR TITLE
[Navigation] Add new indicators to nav items 

### PR DESCRIPTION
**Pull request by dfmcphee**
### WHY are these changes introduced?

Currently we don't have a way to indicate to merchants that an item in the navigation has a sub item that is new.

### WHAT is this pull request doing?

This adds a `new` prop to items and subitems that will callout a new item to the merchant.

#### Example

screen shot 2018-10-26 at 5 27 53 pm

## 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import * as React from 'react';
import {Page, AppProvider, Navigation} from '@shopify/polaris';

interface State {}

export default class Playground extends React.Component<never, State> {
  render() {
    return (
      <AppProvider>
        <Page title="Playground">
          <Navigation
            location="/"
            userMenu={
              <Navigation.UserMenu
                name="Ellen Ochoa"
                detail="Ochoa Crafts"
                actions={[
                  {
                    id: '123',
                    items: [{content: 'Perform action'}],
                  },
                ]}
                message={{
                  title: 'Message title',
                  description: 'Message description',
                  action: {content: 'Perform action'},
                  link: {to: 'path/to/place', content: 'Link content'},
                }}
                avatarInitials="EO"
              />
            }
          >
            <Navigation.Section
              items={[
                {
                  url: '/',
                  label: 'Products',
                  icon: 'products',
                  subNavigationItems: [
                    {
                      url: '/',
                      label: 'All Products',
                    },
                    {
                      url: '/inventory',
                      label: 'Inventory',
                      new: true,
                    },
                    {
                      url: '/collections',
                      label: 'Collections',
                    },
                  ],
                },
              ]}
            />
          </Navigation>
        </Page>
      </AppProvider>
    );
  }
}
```

</details>

### 🎩 checklist

* [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [ ] Updated [UNRELEASED.md](https://github.com/Shopify/polaris-react/blob/master/UNRELEASED.md)

### CODE REVIEW

File: src/components/Navigation/README.md [hunk 1]
```diff
@@ -127,6 +127,7 @@ The content of the navigation component consists of navigation items. Each item
 | badge              | string \| null      | A string property allowing content to be displayed in a badge next to the navigation item                         |
 | label              | string              | A string property allowing content to be displayed as link text in the navigation item                            |
 | disabled           | boolean             | A boolean property indicating whether the navigation item is disabled                                             |
+| new                | boolean             | A boolean property indicating whether the navigation item is new                                                  |
```
Comment (ry5n) [inline src/components/Navigation/README.md]: Should we mention that this adds the indicator dot to the parent? And that it will be overridden by the `badge` prop?

Comment (ry5n) [reply src/components/Navigation/README.md]: (I don’t think the descriptions here need to give the datatype either; there’s a separate column for that)

File: src/components/Navigation/components/Item/Item.tsx [hunk 2]
```diff
@@ -88,20 +99,34 @@ export default class Item extends React.Component<Props, State> {
...
+        .length > 0;
+
+    const indicatorMarkup = hasNewChild ? (
+      <span className={styles.Indicator}>
+        <Indicator pulse />
+      </span>
+    ) : null;
+
+    const badgeMarkup =
+      badge ||
```
Comment (ry5n) [inline src/components/Navigation/components/Item/Item.tsx]: Is this a regression? Looks like `badgeMarkup` now gets the bare value of `badge`. It should still get wrapped in markup.

File: src/components/Navigation/components/Item/Item.tsx [hunk 3]
```diff
@@ -134,6 +159,7 @@ export default class Item extends React.Component<Props, State> {
           >
             {iconMarkup}
             <span className={styles.Text}>{label}</span>
+            {indicatorMarkup}
```
Comment (ry5n) [inline src/components/Navigation/components/Item/Item.tsx]: Is there anything we can do about the indicator wrapping?

File: src/components/Navigation/components/Item/Item.tsx [hunk 4]
```diff
@@ -226,7 +252,10 @@ export default class Item extends React.Component<Props, State> {
           onClick={this.getClickHandler(onClick)}
         >
           {iconMarkup}
-          <span className={styles.Text}>{label}</span>
+          <span className={styles.Text}>
+            {label}
+            {indicatorMarkup}
```
Comment (ry5n) [inline src/components/Navigation/components/Item/Item.tsx]: For the UnstyledLink code path here, the indicator markup is a child of the `.Text` span. For the button path it’s the next sibling. Can they not be done the same way?

File: src/components/Navigation/components/Item/Item.tsx (hunk 2 above)
Comment (dfmcphee) [reply src/components/Navigation/components/Item/Item.tsx]: Oh this is just bad formatting, the bare badge value was just to use as a conditional for the rendering. I will change it to a ternary to match the other conditional markup.

File: src/components/Navigation/components/Item/Item.tsx (hunk 3 above)
Comment (dfmcphee) [reply src/components/Navigation/components/Item/Item.tsx]: Meaning can we prevent it from wrapping while allowing the text to still wrap if needed?

File: src/components/Navigation/components/Item/Item.tsx (hunk 4 above)
Comment (dfmcphee) [reply src/components/Navigation/components/Item/Item.tsx]: Good catch. They should be the same. I moved this content up to a variable defined above so they always stay in sync.

**Comment by dfmcphee** [comment]
Addressed you comments @ry5n. The only outstanding question is how we want to handle the wrapping for the indicator on long labels. I am also ok with deferring this if we aren't sure yet.

File: src/components/Navigation/components/Item/Item.tsx (hunk 3 above)
Comment (ry5n) [reply src/components/Navigation/components/Item/Item.tsx]: Yeah like avoiding the indicator from wrapping below the text, before the text itself wraps. We could try to mitigate this by turning on hyphenation for the text, and having a non-breaking space separate the word and the indicator, but that would need to be tested. I’m also happy to defer this, but it’s worth testing with the menu items we know are going to get this soon, combined with our current set of locales. In other words, it’s OK to defer unless it’s going to immediately look bad to some merchants :)

File: src/components/Navigation/components/Item/Item.tsx [hunk 5]
```diff
@@ -57,7 +68,7 @@ enum MatchState {
   NoMatch,
 }
 
-export default class Item extends React.Component<Props, State> {
+export class BaseItem extends React.Component<CombinedProps, State> {
```
Comment (AndrewMusgrave) [inline src/components/Navigation/components/Item/Item.tsx]: I'm curious, why did you change the class to `BaseItem` and export like 
```tsx
export const Item = withAppProvider<Props>()(BaseItem);
export default Item
```
rather than 
```tsx
export default withAppProvider<Props>(Item)
````

Comment (dfmcphee) [reply src/components/Navigation/components/Item/Item.tsx]: I needed to do this because we use `Item` to [create the sub-navigation](https://github.com/Shopify/polaris-react/blob/49294ee23209ec3a844e916f1a7af4cc524fc933/src/components/Navigation/components/Item/Item.tsx#L226) items and this also needs to be wrapped in `withAppProvider`.

**Comment by tmlayton** [comment]
Want to ship this with v3.0.0?
